src/main/webapp/resources/js/pages/projects/samples-metadata-import2/components/SampleMetadataImportReview.jsx

```diff
@@ -0,0 +1,72 @@
+import React, { useEffect, useState } from "react";
+import { navigate } from "@reach/router"
+import {
+  Button,
+  Space,
+  Table,
+  Tag,
+  Typography,
+} from "antd";
+import { SampleMetadataImportWizard } from "./SampleMetadataImportWizard";
+import { useGetProjectSampleMetadataQuery } from "../../../../apis/metadata/metadata-import";
+import { IconArrowLeft, IconArrowRight } from "../../../../components/icons/Icons";
+
+const { Text } = Typography
+
+/**
+ * React component that displays Step #3 of the Sample Metadata Uploader.
+ * This page is where the user reviews the metadata to be uploaded.
+ * @returns {*}
+ * @constructor
+ */
+export function SampleMetadataImportReview({ projectId }) {
+  const [selected, setSelected] = useState([]);
+  const { data } = useGetProjectSampleMetadataQuery(projectId);
+
+  const dataSource = data?.rows?.map((item, index) => {
+    let newItem = { ...item, key: index };
+    return newItem;
+  });
+
+  const tagColumn = { title: '', dataIndex: 'tags', fixed: 'left', width: 70, render: (text, item) => { if (!item.foundSampleId) return (<Tag color="green">New</Tag>) } };
+
+  const sampleColumn = data?.headers?.filter(item => item === data?.sampleNameColumn).map((header) => {
+    let item = { title: header, dataIndex: header, fixed: 'left', width: 100, render: (text, item) => (<>{item.entry[header]}</>) };
+    return item;
+  });
+
+  const otherColumns = data?.headers?.filter(item => item !== data?.sampleNameColumn).map((header) => {
+    let item = { title: header, dataIndex: header, render: (text, item) => (<>{item.entry[header]}</>) };
+    return item;
+  });
+
+  const columns = sampleColumn?.concat(tagColumn).concat(otherColumns);
+
```

Comment:
Anything that depends on `data` should be put into a `React.useEffect` hook, and stored as state

Example:
```javascript
const [columns, setColumns] = React.useState([]);

React.useEffect(() => {
    if(!isLoading) {
      // Use findIndex that way you don't need to filter the list twice
      const nameIndex = data.headers.findIndex(item => item === data.sampleNameColumn);
      // Remove the sample name data and save it to use later
      const name = data.headers.splice(nameIndex, nameIndex + 1);
      const nameColumn = { 
        title: name, 
        dataIndex: name, 
        fixed: 'left', 
        width: 100, 
        render: (text, item) => (<>{item.entry[name]}</>) 
      }
      
      // Create the rest of the columns
      const columns = data.headers.map(header => ({
        title: header, 
        dataIndex: header, 
        render: (text, item) => (<>{item.entry[header]}</>)}));
      
      // Save the state of the columsn
      setColumns([nameColumn, tagColumn, ...columns])
    }
}, [data, isLoading]);  // isLoading is provided inthe destructure with data
```

I did not test out that code but something like that would be a little better.

Reply:
Please see 913e772da6aa7d88d5ca87de99dc74e0e3ec6cf1.